Encryption/SPLAY/splay.1
.\" File: splay.1
.\" Author: Jeffrey Chilton, PO Box 807, West Branch, IA 52358.
.\" Author: Douglas Jones, Dept. of Comp. Sci., U. of Iowa, Iowa City, IA 52242.
.\" Date: Feb. 20, 1989.
.\" Note: This is not copyrighted.  Documentation should be free!
.\" Language: nroff -man (also troff, eroff ...)
.\" Purpose: UNIX man page for Data compression and encryption program.
.\"
.TH SPLAY 1 "Feb. 20, 1989"
.UC 6
.SH NAME
splay, unsplay \- compress and uncompress data
.SH SYNOPSIS
.PU
.ll +8
.B splay
[
.B \-c
] [
.B \-f
] [
.B \-n
] [
.B \-v
] [
.B \-s
.I states
] [
.I filename
] [
.B \-k
.I keyfile
] [
.B \-p
.I password
]
.ll -8
.br
.B unsplay
[
.B \-c
] [
.B \-f
] [
.B \-n
] [
.I filename
] [
.B \-k
.I keyfile
] [
.B \-p
.I password
]
.SH DESCRIPTION
.I Splay
reduces the size of
.I filename
using prefix codes derived from a binary splay tree.  Whenever possible,
the file is replaced by one with the extension
.B "\&.S."
If no files are specified, the standard input is compressed to the
standard output.  Compressed files with names ending in
.B "\&.S"
can be restored to their original form using
.I unsplay.
.I Splay/unsplay
can also be used as a filter.
.PP
The
.B \-c
(``cat'') option makes
.I splay/unsplay
write to the standard output; no other files are changed or deleted.  If used,
.B \-c
must be specified before all other arguments except
.B \-f, \-n,
and
.B \-v.
.PP
The
.B \-f
option will force the compression or uncompression of the input file,
even if it does not actually shrink, or the corresponding output file
already exists.  If
.B \-f
is not given the user is prompted as to whether an existing
file should be overwritten and the splay command will fail if the
compressed file is larger than the original.  In case of failure,
the output file is deleted and the input file is left unchanged.
.B \-f
must be specified before
.I filename.
.PP
By default,
.I splay/unsplay
will delete the input file.  The
.B \-n
option suppresses this.
.PP
The
.B \-v
option causes the printing of the percentage reduction of the input file.
It also prints the number of bytes used to create the splay tree(s).
.PP
.I Splay
uses the splay prefix algorithm first presented in "Applications
of Splay Trees to Data Compression", D.W. Jones,
.I Communication of the ACM,
August 1988, pages 996-1007.  A balanced splay tree is first constructed by the
program.  As each character is pulled from the input text, it is assigned a
binary prefix code derived by following the path from the root to the proper
leaf.  At each compression, the tree is semi-splayed in such a way
that the distance from the root to the leaf is shortened.
In this way, the most frequently used characters acquire the shortest codes.
.PP
By default,
.I splay/unsplay
uses a Markov compression algorithm which associates a splay tree with each
of 32 states (unless encryption is requested, in which case the default is 1).
The number of states can be changed by the
.B \-s
option, followed by the desired number of states (1 - 256).
When compressing a character, the state used is determined
by the previous character divided by the total number of states.
For 64K machines, no more than 20 states are recommended; for text,
more than 32 states usually leads to little improvement, but about 100 states
is sufficient to compete with the UNIX
.I compress
command.  Digitized images can frequently be compressed better than
.I compress
command with only 1 state.
.PP
.I Splay
can be used for data encryption.
By using the
.B \-p
option, the user may specify a password to be used as an encryption key.
The
.B \-k
option may be specified to enable the program to use a file to hold the
encryption key.  Before compressing or uncompressing the input, the initial
tree used for data compression is semi-splayed about each byte of the key,
thus making it very difficult to recover the compressed data without
knowledge of the key.  Any number of
.B \-p
or
.B \-k
options may be used; if multiple keys are specified, they will be
concatenated in the order specified.
The
.B \-p
and
.B \-k
options must be specified after the filename, if any is given.
.PP
Any time an encryption key is specified,
.I splay
prefixes several random bytes to the source file in order to complicate
known-text attacks.  These bytes are automatically discarded by
.I unsplay.
If a key is specified,
.I splay
will retain the output even if it is larger than the input.
.PP
.SH "DIAGNOSTICS"
Error: bad
.B \-x
argument
Usage:
.I splay/unsplay
[\-c] [\-f] [\-n] [\-v] [\-s states] [filename] [\-k keyfile] [\-p password]
.in +8
Invalid options were specified on the command line.
-c must precede all other arguments.
Keys must be specified after filename.
-s must be specified before keys.
.in -8
Error:
.IR splay/unsplay
can't open
.IR filename
.in +8
Unable to open this file.  File does not exist, or is inaccessible.
.in -8
Error:
.IR filename
is not in splayed format
.in +8
.I Unsplay
cannot unsplay the file, because it has not been splayed.
All splayed files are given the magic number ^S^P.
.in -8
Error: corrupt input in
.IR filename
.in +8
Either the file has been encrypted and the proper decoding key has
not been entered, or the splayed data has been changed or ruined.
.in -8
Error: cannot
.IR splay/unsplay
.IR filename: .S
suffix expected
.in +8
Input file is required to have a .S suffix for unsplay to be able to use it.
.in -8
Error:
.IR filename
could not be compressed
.in +8
.I Splay
was unable to produce a smaller file.
.in -8
Error: too many arguments
.in +8
More than one filename has been specified or number of states specified is in
conflict with a previous setting.
.in -8
Error: \-s
.B x
not in bounds of 1 to 256
.in +8
The number of states is not within the range 1 - 256.
.in -8
.SH "SEE ALSO"
.br
crypt(1), compress(1), uncompress(1), zcat(1)
.SH BUGS
As with most encryption algorithms, we cannot ensure the security of the
encrypted result.  As of this writing (February 20, 1989), we know nothing
about deciphering encrypted text produced using splay trees.  In the short
run, this may provide a small degree of security, but in the long run, it
should only reinforce the insecurity of using this method.
.PP
For all of the text we have compressed and uncompressed, the output of the
.I unsplay
program has been the same as the original input to the
.I splay
program, but we cannot warrant that this will be the case for files
we have not tried.
.PP
Unlike
.I compress
and
.I uncompress,
this package makes no effort to preserve the ownership modes, access times,
and modification times of the files processed.
This package will compress or uncompress only one file at a time.
.SH AUTHORS
Douglas W. Jones and Jeffrey E. Chilton


Encryption/SPLAY/splay.c
/* File: splay.c
   Author: Jeffrey Chilton, PO Box 807, West Branch, IA 52358.
   Author: Douglas Jones, Dept. of Comp. Sci., U. of Iowa, Iowa City, IA 52242.
   Date: Feb. 14, 1990
	 (minor revision of Feb. 20, 1989 to add exit(0) at end of program).
	 (minor revision of Nov. 14, 1988 to fix portability problems).
	 (minor revision of Aug. 8, 1988 to eliminate unused vars, fix -c).
   Copyright 1988 by Jeffrey Chilton and Douglas Jones.
	      Copies of this program and associated files may not be sold,
			nor may it (or parts of it) be incorporated into
			products which will be sold, without the express
			prior permission of the authors.
	      Copies of this program, associated files, or parts thereof may
			not be made for use by any national government or
			agency thereof without the express prior permission
			of the authors.
	      Permission is hereby granted to make copies of this program for
			research use (including research in commercial or
			governmental settings) so long as this copyright
			notice is included in the copy and in any derived work.
   Language: C (UNIX)
   Purpose: Data compression and encryption program
   Algorithm: Uses a splay-tree based prefix code, with one splay tree per
			state in a Markov model.  The nature of the Markov
			model is determined by the splay procedure in the
			include file splay.i.
*/

#include <stdio.h>
#include <strings.h>
#include <time.h>
#include "splay.i"

/* begin TRANSMIT macro */
#define MAXBITCOUNTER 8
short int bitbuffer = 0;
short int bitcounter = 0;
#define TRANSMIT(b) {						\
	bitbuffer = (bitbuffer << 1) + (b);			\
	if ((++bitcounter) == MAXBITCOUNTER) {			\
		putc(bitbuffer, out);				\
		bitcounter = bitbuffer = 0;			\
	}							\
}
/* end TRANSMIT macro */

short int stack[SUCCMAX]; /* used by compress, statically allocated for speed */

/* begin COMPRESS macro */
/* COMPRESS macro is used where speed is essential, and if your compiler */
/* can't handle the macro, or if you are out of memory, use the function */
/* compress and delete the macro */
#define COMPRESS(plain) {					\
	short int *U, *R;					\
	short int sp = 0;					\
	short int a = (plain) + SUCCMAX;			\
	R = &right[state * SUCCMAX];				\
	U = &up[state * SUCCTWICE];				\
	do { 		/* walk up the tree pushing bits */	\
		stack[sp] = R[U[a]] == a;			\
		++sp;						\
		a = U[a];					\
 	} while (a != ROOT);					\
	do {							\
		TRANSMIT(stack[(--sp)]);			\
	} while (sp != 0);					\
	SPLAY((short int)(plain));				\
}
/* end COMPRESS macro */

/* begin compress function */
compress(plain)
short int plain;
{
	COMPRESS(plain);
}
/* end compress function */

/* begin openfiles function */
openfiles()
{
	if (filename == NULL) {
		in = stdin;
		out = stdout;
		cflag = 1;
		compflag = 1;
		rmfileflag = 0;
	} else {
		if ((in = fopen(filename, "r")) == NULL) cannot_open(filename);
		if (cflag == 0) {
			strncpy(filenmod, filename, 117);
			strncat(filenmod, ".S", 3);
			if ((out = fopen(filenmod, "a")) == NULL) {
				cannot_open(filenmod);
			}
			if ((forceflag == 0) && (ftell(out) != 0)) overwrite();
		} else out = stdout;
	}
}
/* end openfiles function */

/* THE MAIN PROGRAM */
main(argc, argv)
int argc;
char *argv[];
{
	int plain;
	long int oldlen, newlen;

	linearg(argc, argv);

	if (states == 0) {
		states = nokeydefault;
		initsplay();
	}

	putc(MAGIC1, out);
	putc(MAGIC2, out);
	if (states == 256) {
		putc(NULL, out);
	} else {
		putc((char)states, out);
	}

	/* begin transmit garbage prefix */
	if (cryptflag == 1) {
		long int pid, t, k;
		pid = getpid();
		t = time(NULL);
		k = (pid & 0x00007fffL) * (t & 0x00007fffL);
		compress((short)k & 0xff);
		compress((short)(k >> 8) & 0xff);
		compress((short)(k >> 16) & 0xff);
		compress((short)(k >> 24) & 0xff);
	}
	/* end transmit garbage prefix */

	/* begin compress input */
	for (;;) {
		if (((plain = getc(in)) == EOF) && feof(in)) break;
		COMPRESS ((short int)(plain & 0xff));
	}
	compress(MAXCHAR);
	/* end compress input */

	/* begin flushtransmit routine */
	while (bitcounter != 0) {
		TRANSMIT(0);
	}
	/* end flushtransmit routine */

	/* begin calculate compression */
	oldlen = ftell(in);
	newlen = ftell(out);
	if (vflag == 1) {
		int bytes = 2 * (states * SUCCMAX * sizeof(short)) + (states *
			SUCCTWICE * sizeof(short));
		float comp = 100.0 * (1.0 - ((float)newlen / (float)oldlen));
		fprintf(stderr, "%s: Compression: %2.1f%% ", filename, comp);
		if (rmfileflag == 1) {
			fprintf(stderr, "-- replaced with %s\n", filenmod);
		} else fprintf(stderr, "\n");
		fprintf(stderr, "Data structure used %d bytes\n", bytes);
	}
	/* end calculate compression */

	if ((newlen > oldlen) && (compflag == 0)) {
		unlink(filenmod);
		already_compressed();
	}
	if ((filename != NULL) && (rmfileflag == 1)) unlink(filename);
	exit(0);
}
